### PR TITLE
Agentless EKS example demo and module changes

### PR DESCRIPTION
**Pull request by riddhi89**
Consul 1.14.0 release supports connecting kubernetes workloads without Consul client agents, to HCP Consul servers. In fact since consul helm chart 1.0.0 release, service mesh deployments use [Consul Dataplanes ](https://developer.hashicorp.com/consul/docs/connect/dataplane)(and NO consul client agents) by default.

This PR:

- Updates the hcp-eks-client module to support deploying consul on EKS without client agents, using consul dataplane and also makes this the default deployment model.
- Updates the hcp-eks-demo example to deploy the hashicups application using Agentless (without client agents) Consul service mesh on an EKS cluster, connected to a 1.14 HCP Consul server.

### CODE REVIEW

File: examples/hcp-eks-demo/main.tf [hunk 1]
```diff
@@ -35,7 +35,7 @@ module "eks" {
   version                = "17.24.0"
   kubeconfig_api_version = "client.authentication.k8s.io/v1beta1"
 
-  cluster_name    = "${var.cluster_id}-eks"
+  cluster_name    = "riddhi-aws-114beta-eks"
```
Comment (riddhi89) [inline examples/hcp-eks-demo/main.tf:38]: note-to-self: Temp for testing: revert this change

File: examples/hcp-eks-demo/main.tf [hunk 2]
```diff
@@ -54,46 +54,54 @@ module "eks" {
...
+# resource "hcp_hvn" "main" {
+#   hvn_id         = var.hvn_id
+#   cloud_provider = "aws"
+#   region         = var.hvn_region
+#   cidr_block     = var.hvn_cidr_block
+# }
+
+data "hcp_hvn" "example" {
+  hvn_id = "hvn"
 }
```
Comment (riddhi89) [inline examples/hcp-eks-demo/main.tf:66]: note-to-self: data source for HVN instead of creating HVN - temp for testing: revert this change

File: examples/hcp-eks-demo/main.tf [hunk 3]
```diff
@@ -54,46 +54,54 @@ module "eks" {
...
+#   cluster_id      = var.cluster_id
+#   hvn_id          = hcp_hvn.main.hvn_id
+#   public_endpoint = true
+#   tier            = var.tier
+#   min_consul_version = "v1.14.0"
+# }
+
+data "hcp_consul_cluster"  "main" {
+  cluster_id = "riddhi-aws-114beta"
 }
```
Comment (riddhi89) [inline examples/hcp-eks-demo/main.tf:89]: note-to-self: using data source for 1.14.0 beta cluster instead of creating cluster (cannot create beta clusters via terraform) temp for testing: revert this change

File: examples/hcp-eks-demo/main.tf [hunk 4]
```diff
@@ -54,46 +54,54 @@ module "eks" {
...
-  cluster_id = hcp_consul_cluster.main.id
+  cluster_id = data.hcp_consul_cluster.main.id
 }
 
 module "eks_consul_client" {
-  source  = "hashicorp/hcp-consul/aws//modules/hcp-eks-client"
-  version = "~> 0.8.9"
+  # source  = "hashicorp/hcp-consul/aws//modules/hcp-eks-client"
+  # version = "~> 0.8.9"
+  source = "../../modules/hcp-eks-client/"
```
Comment (riddhi89) [inline examples/hcp-eks-demo/main.tf:98]: note-to-self: Temp for testing: revert this change

File: examples/hcp-eks-demo/output.tf [hunk 5]
```diff
@@ -4,10 +4,10 @@ output "consul_root_token" {
 }
 
 output "consul_url" {
-  value = hcp_consul_cluster.main.public_endpoint ? (
-    hcp_consul_cluster.main.consul_public_endpoint_url
+  value = data.hcp_consul_cluster.main.public_endpoint ? (
+    data.hcp_consul_cluster.main.consul_public_endpoint_url
     ) : (
-    hcp_consul_cluster.main.consul_private_endpoint_url
+    data.hcp_consul_cluster.main.consul_private_endpoint_url
```
Comment (riddhi89) [inline examples/hcp-eks-demo/output.tf:10]: note-to-self: Temp for testing: revert this change

File: examples/hcp-eks-demo/output.tf [hunk 6]
```diff
@@ -33,7 +33,7 @@ output "howto_connect" {
   ${var.install_demo_app ? "To access HashiCups navigate to: ${module.demo_app[0].hashicups_url}" : ""}
 
   To access Consul from your local client run:
-  export CONSUL_HTTP_ADDR="${hcp_consul_cluster.main.consul_public_endpoint_url}"
+  export CONSUL_HTTP_ADDR="${data.hcp_consul_cluster.main.consul_public_endpoint_url}"
```
Comment (riddhi89) [inline examples/hcp-eks-demo/output.tf:36]: note-to-self: Temp for testing: revert this change

File: modules/hcp-eks-client/templates/consul.tpl [hunk 7]
```diff
@@ -28,17 +37,16 @@ externalServers:
...
-client:
-  enabled: true
-  join: ${consul_hosts}
-  nodeMeta:
-    terraform-module: "hcp-eks-client"
-
 connectInject:
   transparentProxy:
-    defaultEnabled: true
+    defaultEnabled: false
```
Comment (riddhi89) [inline modules/hcp-eks-client/templates/consul.tpl:42]: note-to-self: change this back to true for GA (Tproxy not supported for beta)

File: modules/hcp-eks-client/templates/consul.tpl [hunk 8]
```diff
@@ -28,17 +37,16 @@ externalServers:
...
   enabled: true
   default: true
+%{ if !consul_client_agent ~}
+  consulNode:
+    meta: "hcp-eks-agentless-node"
+%{ endif ~}
+
```
Comment (riddhi89) [inline modules/hcp-eks-client/templates/consul.tpl:49]: validate this before GA: required for HCP analytics (currently nothing happens if this is included as consul-k8s does not support it yet)

File: examples/hcp-eks-demo/main.tf [hunk 9]
```diff
@@ -54,46 +54,54 @@ module "eks" {
...
+# Note: Uncomment the below module to setup peering for connecting to a private HCP Consul cluster
+# module "aws_hcp_consul" {
+#   source  = "hashicorp/hcp-consul/aws"
+#   version = "~> 0.8.9"
+#   hvn                = hcp_hvn.main
+#   vpc_id             = module.vpc.vpc_id
+#   subnet_ids         = module.vpc.private_subnets
+#   route_table_ids    = module.vpc.private_route_table_ids
+#   security_group_ids = var.install_eks_cluster ? [module.eks[0].cluster_primary_security_group_id] : [""]
+# }
```
Comment (riddhi89) [inline examples/hcp-eks-demo/main.tf:77]: Note: This will remain commented to stay as a reference in case the user wants to setup peering. We could argue that the top level module is clear enough - so open to feedback if we want to remove this.

Comment (crhino) [reply examples/hcp-eks-demo/main.tf:77]: I think it is reasonable to leave for now. 👍

File: scripts/eks_existing_vpc_locals.snip [hunk 10]
```diff
@@ -1,5 +1,4 @@
   vpc_id                 = "{{ .VPCID }}"
-  private_route_table_id = "{{ .PrivateRouteTableID }}"
```
Comment (crhino) [inline scripts/eks_existing_vpc_locals.snip:2]: Why did this change?

File: modules/hcp-eks-client/variables.tf [hunk 11]
```diff
@@ -48,12 +38,24 @@ variable "consul_version" {
...
  * Optional Variables
  *
- */
+*/
 
 variable "chart_version" {
   type        = string
   description = "The Consul Helm chart version to use"
-  default     = "0.41.1"
+  default     = "1.0.0-beta3"
```
Comment (crhino) [inline modules/hcp-eks-client/variables.tf:46]: Note: should update this when 1.0.0 is actually released.

File: scripts/eks_existing_vpc_locals.snip (hunk 10 above)
Comment (riddhi89) [reply scripts/eks_existing_vpc_locals.snip:2]: This field is only needed to setup peering. On the UI, we were considering the HCP AWS -> EKS -> Existing VPC flow to create public 1.14.0 clusters by default and not setup peering. Hence, removed this.

File: examples/hcp-eks-demo/main.tf [hunk 12]
```diff
@@ -54,46 +54,55 @@ module "eks" {
...
+
+# resource "hcp_consul_cluster" "main" {
+#   cluster_id      = var.cluster_id
+#   hvn_id          = hcp_hvn.main.hvn_id
+#   public_endpoint = true
+#   tier            = var.tier
+#   min_consul_version = "v1.14.0"
```
Comment (riddhi89) [inline examples/hcp-eks-demo/main.tf:84]: For the example, clusters will be created at 1.14.0 by default (to not need peering).

File: modules/hcp-eks-client/templates/consul.tpl [hunk 13]
```diff
@@ -28,17 +37,17 @@ externalServers:
...
   enabled: true
   default: true
+%{ if !consul_client_agent ~}
+  consulNode:
+    meta:
+      terraform-module: "hcp-eks-client"
+%{ endif ~}
+
 
 controller:
```
Comment (riddhi89) [inline modules/hcp-eks-client/templates/consul.tpl:52]: note-to-self: remove this for agentless (connectInject.enabled = true should suffice).

File: examples/hcp-eks-demo/main.tf [hunk 14]
```diff
@@ -61,22 +62,24 @@ resource "hcp_hvn" "main" {
...
 resource "hcp_consul_cluster" "main" {
-  cluster_id      = var.cluster_id
-  hvn_id          = hcp_hvn.main.hvn_id
-  public_endpoint = true
-  tier            = var.tier
+  cluster_id         = var.cluster_id
+  hvn_id             = hcp_hvn.main.hvn_id
+  public_endpoint    = true
+  tier               = var.tier
+  min_consul_version = "v1.14.0"
```
Comment (riddhi89) [inline examples/hcp-eks-demo/main.tf:82]: By default, example uses a 1.14.0 cluster to demo an Agentless deployment.

File: examples/hcp-eks-demo/providers.tf [hunk 15]
```diff
@@ -12,17 +12,17 @@ terraform {
 
     kubernetes = {
       source  = "hashicorp/kubernetes"
-      version = ">= 2.4.1"
+      version = "~> 2.14.0"
```
Comment (riddhi89) [inline examples/hcp-eks-demo/providers.tf:15]: Pinned the provider versions for better predictable deployments.

File: modules/k8s-demo-app/services/nginx.yaml [hunk 16]
```diff
@@ -46,16 +46,16 @@ data:
         proxy_buffering off;
 
         location / {
-          proxy_pass http://127.0.0.1:3000;
+          proxy_pass http://frontend:3000;
```
Comment (riddhi89) [inline modules/k8s-demo-app/services/nginx.yaml:49]: Can do this since transparent proxy is enabled.

**Review by crhino — APPROVED** [review]
LGTM!
